// @saberhq/solana-contrib/dist/cjs/broadcaster/tiered.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TieredBroadcaster = void 0;
const provider_js_1 = require("../provider.js");
const index_js_1 = require("../transaction/index.js");
const index_js_2 = require("../utils/index.js");
const index_js_3 = require("./index.js");
const sendAndSpamRawTx_js_1 = require("./sendAndSpamRawTx.js");
/**
 * Broadcasts transactions to multiple connections simultaneously.
 */
class TieredBroadcaster {
    constructor(primaryConnection, 
    /**
     * Connections to send to in addition to the primary.
     */
    fallbackConnections, opts = provider_js_1.DEFAULT_PROVIDER_OPTIONS) {
        this.primaryConnection = primaryConnection;
        this.fallbackConnections = fallbackConnections;
        this.opts = opts;
        this.premiumBroadcaster = new index_js_3.SingleConnectionBroadcaster(primaryConnection, opts);
    }
    async getLatestBlockhash(commitment) {
        var _a;
        if (commitment === void 0) { commitment = (_a = this.opts.preflightCommitment) !== null && _a !== void 0 ? _a : "confirmed"; }
        return await this.premiumBroadcaster.getLatestBlockhash(commitment);
    }
    async getRecentBlockhash(commitment) {
        var _a;
        if (commitment === void 0) { commitment = (_a = this.opts.preflightCommitment) !== null && _a !== void 0 ? _a : "confirmed"; }
        return await this.premiumBroadcaster.getRecentBlockhash(commitment);
    }
    async _sendRawTransaction(encoded, options) {
        const pending = new index_js_1.PendingTransaction(this.primaryConnection, await (0, sendAndSpamRawTx_js_1.sendAndSpamRawTx)(this.primaryConnection, encoded, options !== null && options !== void 0 ? options : this.opts, options !== null && options !== void 0 ? options : index_js_3.DEFAULT_RETRY_OPTIONS));
        void (async () => {
            await Promise.all(this.fallbackConnections.map(async (fc) => {
                var _a;
                await (0, sendAndSpamRawTx_js_1.sendAndSpamRawTx)(fc, encoded, options !== null && options !== void 0 ? options : this.opts, (_a = options === null || options === void 0 ? void 0 : options.fallbackRetryOptions) !== null && _a !== void 0 ? _a : index_js_3.DEFAULT_FALLBACK_RETRY_OPTIONS);
            }));
        })();
        return pending;
    }
    /**
     * Broadcasts a signed transaction.
     *
     * @param tx
     * @param confirm
     * @param opts
     * @returns
     */
    async broadcast(tx, { printLogs = true, ...opts } = this.opts) {
        if (tx.signatures.length === 0) {
            throw new Error("Transaction must be signed before broadcasting.");
        }
        const rawTx = tx.serialize();
        if (printLogs) {
            return await this._sendRawTransaction(rawTx, opts);
        }
        return await (0, index_js_2.suppressConsoleErrorAsync)(async () => {
            // hide the logs of TX errors if printLogs = false
            return await this._sendRawTransaction(rawTx, opts);
        });
    }
    /**
     * Simulates a transaction with a commitment.
     * @param tx
     * @param commitment
     * @returns
     */
    async simulate(tx, _a) {
        var _b, _c;
        var { commitment = (_b = this.opts.preflightCommitment) !== null && _b !== void 0 ? _b : "confirmed", verifySigners = true, } = _a === void 0 ? {
            commitment: (_c = this.opts.preflightCommitment) !== null && _c !== void 0 ? _c : "confirmed",
            verifySigners: true,
        } : _a;
        if (verifySigners && tx.signatures.length === 0) {
            throw new Error("Transaction must be signed before simulating.");
        }
        return this.premiumBroadcaster.simulate(tx, {
            commitment,
            verifySigners,
        });
    }
}
exports.TieredBroadcaster = TieredBroadcaster;
