// @saberhq/anchor-contrib/dist/esm/utils/coder.js
import { BorshCoder, EventParser, utils } from "@project-serum/anchor";
import { default as InstructionNamespaceFactory } from "@project-serum/anchor/dist/cjs/program/namespace/instruction.js";
import { TransactionInstruction } from "@solana/web3.js";
import mapValues from "lodash.mapvalues";
import { generateErrorMap } from "../errors.js";
import { generateAccountParsersFromCoder } from "../generateAccountParsers.js";
import { generateAnchorAccounts } from "./accounts.js";
import { newProgram } from "./programs.js";
/**
 * Coder wrapper.
 *
 * Allows interacting with a program without a provider.
 */
export class SuperCoder {
    /**
     * Constructor.
     * @param address
     * @param idl
     */
    constructor(
    /**
     * Program address.
     */
    address, 
    /**
     * Program IDL.
     */
    idl) {
        var _a, _b;
        this.address = address;
        this.idl = idl;
        this.coder = new BorshCoder(idl);
        this.eventParser = new EventParser(address, this.coder);
        this.accounts = generateAnchorAccounts(address, (_a = idl.accounts) !== null && _a !== void 0 ? _a : [], this.coder.accounts);
        this.errorMap = generateErrorMap(idl);
        const accountsList = Object.values(this.accounts);
        const accountTypeDefs = {};
        accountsList.forEach((account) => {
            accountTypeDefs[account.name] = account.idl;
        });
        this.accountParsers = generateAccountParsersFromCoder((_b = idl.accounts) === null || _b === void 0 ? void 0 : _b.map((acc) => acc.name), this.coder.accounts);
        this.accountTypeDefs = accountTypeDefs;
        this.discriminators = accountsList.reduce((acc, el) => ({ ...acc, [el.discriminator.toString("hex")]: el.name }), {});
        this.discriminatorsByAccount = accountsList.reduce((acc, el) => ({ ...acc, [el.name]: el.discriminator }), {});
    }
    /**
     * Creates a {@link GetProgramAccountsFilter} for the given account.
     */
    makeGPAFilter(account, ...filters) {
        return [
            {
                memcmp: {
                    offset: 0,
                    bytes: utils.bytes.bs58.encode(this.discriminatorsByAccount[account]),
                },
            },
            ...filters,
        ];
    }
    /**
     * Parses events in the program log.
     * @param logs
     * @returns
     */
    parseProgramLogEvents(logs) {
        if (!logs) {
            return [];
        }
        const events = [];
        this.eventParser.parseLogs(logs !== null && logs !== void 0 ? logs : [], (event) => events.push(event));
        return events;
    }
    /**
     * Encodes a {@link TransactionInstruction}.
     * @returns
     */
    encodeIX(name, args, accounts) {
        const idlIx = this.idl.instructions.find((ix) => ix.name === name);
        if (!idlIx) {
            throw new Error(`could not find ix: ${name}`);
        }
        const encoded = this.coder.instruction.encode(name, args);
        const keys = InstructionNamespaceFactory.accountsArray(accounts, idlIx.accounts, name);
        return new TransactionInstruction({
            programId: this.address,
            keys,
            data: encoded,
        });
    }
    /**
     * Parses a {@link TransactionInstruction}.
     * @returns
     */
    parseInstruction(txInstruction) {
        const decoded = this.coder.instruction.decode(txInstruction.data);
        if (!decoded) {
            throw new Error("could not decode ix data");
        }
        const fmt = this.coder.instruction.format(decoded, txInstruction.keys);
        if (!fmt) {
            throw new Error("invalid instruction");
        }
        return { ...fmt, name: decoded.name };
    }
    /**
     * Gets a {@link Program} from a provider.
     * @param provider
     * @returns
     */
    getProgram(provider) {
        return newProgram(this.idl, this.address, provider);
    }
}
/**
 * Builds a map of coders from their IDLs and addresses.
 *
 * @param provider
 * @param programs
 * @returns
 */
export const buildCoderMap = (idls, addresses) => {
    return mapValues(idls, (idl, k) => new SuperCoder(addresses[k], idl));
};
