// @cardinal/stake-pool/dist/cjs/constants.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.STAKE_POOL_IDL = exports.STAKER_OFFSET = exports.GROUP_ENTRY_SEED = exports.STAKE_ENTRY_SEED = exports.STAKE_POOL_PROGRAM_ID = exports.CRANK_PUB_KEY = void 0;
const STAKE_POOL_TYPES = __importStar(require("./idl/stake_pool"));
const web3_js_1 = require("@solana/web3.js");
exports.CRANK_PUB_KEY = new web3_js_1.PublicKey("crkdpVWjHWdggGgBuSyAqSmZUmAjYLzD435tcLDRLXr");
exports.STAKE_POOL_PROGRAM_ID = new web3_js_1.PublicKey("LVLYTWmTaRCV5JcZ5HQkU1bhEjx34xqGiT3eWU6SuX9");
exports.STAKE_ENTRY_SEED = "stake-entry";
exports.GROUP_ENTRY_SEED = "group-stake-entry";
exports.STAKER_OFFSET = 97;
exports.STAKE_POOL_IDL = STAKE_POOL_TYPES.IDL;
