// @cardinal/common/dist/esm/mints.js
import { __awaiter } from "tslib";
import * as splToken from "@solana/spl-token";
import { SystemProgram } from "@solana/web3.js";
import { withFindOrInitAssociatedTokenAccount } from "./transactions";
export const withCreateMint = (transaction, connection, wallet, recipient, mintId, amount = 1, freezeAuthority = recipient) => __awaiter(void 0, void 0, void 0, function* () {
    const mintBalanceNeeded = yield splToken.Token.getMinBalanceRentForExemptMint(connection);
    transaction.add(SystemProgram.createAccount({
        fromPubkey: wallet.publicKey,
        newAccountPubkey: mintId,
        lamports: mintBalanceNeeded,
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
        space: splToken.MintLayout.span,
        programId: splToken.TOKEN_PROGRAM_ID,
    }));
    transaction.add(splToken.Token.createInitMintInstruction(splToken.TOKEN_PROGRAM_ID, mintId, 0, wallet.publicKey, freezeAuthority));
    const walletAta = yield withFindOrInitAssociatedTokenAccount(transaction, connection, mintId, wallet.publicKey, wallet.publicKey);
    if (amount > 0) {
        transaction.add(splToken.Token.createMintToInstruction(splToken.TOKEN_PROGRAM_ID, mintId, walletAta, wallet.publicKey, [], amount));
    }
    return [walletAta, transaction];
});
/**
 * Creates mint, token account, and mints amount to token account
 * @param connection
 * @param creator
 * @returns
 */
export const createMint = (connection, creator, recipient, amount = 1, freezeAuthority = recipient) => __awaiter(void 0, void 0, void 0, function* () {
    const mint = yield splToken.Token.createMint(connection, creator, creator.publicKey, freezeAuthority, 0, splToken.TOKEN_PROGRAM_ID);
    const tokenAccount = yield mint.createAssociatedTokenAccount(recipient);
    yield mint.mintTo(tokenAccount, creator.publicKey, [], amount);
    return [tokenAccount, mint];
});
