// @cardinal/common/dist/cjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.connectionFor = void 0;
const tslib_1 = require("tslib");
// import type { Wallet } from "@saberhq/solana-contrib";
// import type { WalletContextState } from "@solana/wallet-adapter-react";
const web3 = tslib_1.__importStar(require("@solana/web3.js"));
/**
 * Wrap WalletContextState interface into Wallet interface
 * @param wallet
 * @returns
 */
// export const asWallet = (wallet: WalletContextState): Wallet => {
//   return wallet as Wallet;
// };
const networkURLs = {
    ["mainnet-beta"]: "https://ssc-dao.genesysgo.net/",
    mainnet: "https://ssc-dao.genesysgo.net/",
    devnet: "https://psytrbhymqlkfrhudd.dev.genesysgo.net:8899/",
    testnet: "https://api.testnet.solana.com/",
    localnet: "http://localhost:8899/",
};
/**
 * Create and return new Connection object
 * @param cluster
 * @param defaultCluster
 * @returns
 */
const connectionFor = (cluster, defaultCluster = "mainnet") => {
    return new web3.Connection(process.env.RPC_URL || networkURLs[cluster || defaultCluster], "recent");
};
exports.connectionFor = connectionFor;
