// @saberhq/anchor-contrib/dist/cjs/utils/coder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildCoderMap = exports.SuperCoder = void 0;
const tslib_1 = require("tslib");
const anchor_1 = require("@project-serum/anchor");
const instruction_js_1 = tslib_1.__importDefault(require("@project-serum/anchor/dist/cjs/program/namespace/instruction.js"));
const web3_js_1 = require("@solana/web3.js");
const lodash_mapvalues_1 = tslib_1.__importDefault(require("lodash.mapvalues"));
const errors_js_1 = require("../errors.js");
const generateAccountParsers_js_1 = require("../generateAccountParsers.js");
const accounts_js_1 = require("./accounts.js");
const programs_js_1 = require("./programs.js");
/**
 * Coder wrapper.
 *
 * Allows interacting with a program without a provider.
 */
class SuperCoder {
    /**
     * Constructor.
     * @param address
     * @param idl
     */
    constructor(
    /**
     * Program address.
     */
    address, 
    /**
     * Program IDL.
     */
    idl) {
        var _a, _b;
        this.address = address;
        this.idl = idl;
        this.coder = new anchor_1.BorshCoder(idl);
        this.eventParser = new anchor_1.EventParser(address, this.coder);
        this.accounts = (0, accounts_js_1.generateAnchorAccounts)(address, (_a = idl.accounts) !== null && _a !== void 0 ? _a : [], this.coder.accounts);
        this.errorMap = (0, errors_js_1.generateErrorMap)(idl);
        const accountsList = Object.values(this.accounts);
        const accountTypeDefs = {};
        accountsList.forEach((account) => {
            accountTypeDefs[account.name] = account.idl;
        });
        this.accountParsers = (0, generateAccountParsers_js_1.generateAccountParsersFromCoder)((_b = idl.accounts) === null || _b === void 0 ? void 0 : _b.map((acc) => acc.name), this.coder.accounts);
        this.accountTypeDefs = accountTypeDefs;
        this.discriminators = accountsList.reduce((acc, el) => ({ ...acc, [el.discriminator.toString("hex")]: el.name }), {});
        this.discriminatorsByAccount = accountsList.reduce((acc, el) => ({ ...acc, [el.name]: el.discriminator }), {});
    }
    /**
     * Creates a {@link GetProgramAccountsFilter} for the given account.
     */
    makeGPAFilter(account, ...filters) {
        return [
            {
                memcmp: {
                    offset: 0,
                    bytes: anchor_1.utils.bytes.bs58.encode(this.discriminatorsByAccount[account]),
                },
            },
            ...filters,
        ];
    }
    /**
     * Parses events in the program log.
     * @param logs
     * @returns
     */
    parseProgramLogEvents(logs) {
        if (!logs) {
            return [];
        }
        const events = [];
        this.eventParser.parseLogs(logs !== null && logs !== void 0 ? logs : [], (event) => events.push(event));
        return events;
    }
    /**
     * Encodes a {@link TransactionInstruction}.
     * @returns
     */
    encodeIX(name, args, accounts) {
        const idlIx = this.idl.instructions.find((ix) => ix.name === name);
        if (!idlIx) {
            throw new Error(`could not find ix: ${name}`);
        }
        const encoded = this.coder.instruction.encode(name, args);
        const keys = instruction_js_1.default.accountsArray(accounts, idlIx.accounts, name);
        return new web3_js_1.TransactionInstruction({
            programId: this.address,
            keys,
            data: encoded,
        });
    }
    /**
     * Parses a {@link TransactionInstruction}.
     * @returns
     */
    parseInstruction(txInstruction) {
        const decoded = this.coder.instruction.decode(txInstruction.data);
        if (!decoded) {
            throw new Error("could not decode ix data");
        }
        const fmt = this.coder.instruction.format(decoded, txInstruction.keys);
        if (!fmt) {
            throw new Error("invalid instruction");
        }
        return { ...fmt, name: decoded.name };
    }
    /**
     * Gets a {@link Program} from a provider.
     * @param provider
     * @returns
     */
    getProgram(provider) {
        return (0, programs_js_1.newProgram)(this.idl, this.address, provider);
    }
}
exports.SuperCoder = SuperCoder;
/**
 * Builds a map of coders from their IDLs and addresses.
 *
 * @param provider
 * @param programs
 * @returns
 */
const buildCoderMap = (idls, addresses) => {
    return (0, lodash_mapvalues_1.default)(idls, (idl, k) => new SuperCoder(addresses[k], idl));
};
exports.buildCoderMap = buildCoderMap;
