// @cardinal/token-manager/dist/esm/programs/paymentManager/instruction.js
import { AnchorProvider, Program } from "@project-serum/anchor";
import { TOKEN_PROGRAM_ID } from "@solana/spl-token";
import { SystemProgram } from "@solana/web3.js";
import { CRANK_KEY } from "../tokenManager";
import { PAYMENT_MANAGER_ADDRESS, PAYMENT_MANAGER_IDL } from ".";
import { findPaymentManagerAddress } from "./pda";
export const init = async (connection, wallet, name, params) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const paymentManagerProgram = new Program(PAYMENT_MANAGER_IDL, PAYMENT_MANAGER_ADDRESS, provider);
    const [paymentManagerId] = await findPaymentManagerAddress(name);
    return [
        paymentManagerProgram.instruction.init({
            name: name,
            feeCollector: params.feeCollector,
            makerFeeBasisPoints: params.makerFeeBasisPoints,
            takerFeeBasisPoints: params.takerFeeBasisPoints,
        }, {
            accounts: {
                paymentManager: paymentManagerId,
                authority: params.authority || wallet.publicKey,
                payer: wallet.publicKey,
                systemProgram: SystemProgram.programId,
            },
        }),
        paymentManagerId,
    ];
};
export const managePayment = async (connection, wallet, name, params) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const paymentManagerProgram = new Program(PAYMENT_MANAGER_IDL, PAYMENT_MANAGER_ADDRESS, provider);
    const [paymentManagerId] = await findPaymentManagerAddress(name);
    return paymentManagerProgram.instruction.managePayment(params.paymentAmount, {
        accounts: {
            paymentManager: paymentManagerId,
            payerTokenAccount: params.payerTokenAccount,
            feeCollectorTokenAccount: params.feeCollectorTokenAccount,
            paymentTokenAccount: params.paymentTokenAccount,
            payer: wallet.publicKey,
            tokenProgram: TOKEN_PROGRAM_ID,
        },
    });
};
export const close = async (connection, wallet, name, collector) => {
    const provider = new AnchorProvider(connection, wallet, {});
    const paymentManagerProgram = new Program(PAYMENT_MANAGER_IDL, PAYMENT_MANAGER_ADDRESS, provider);
    const [paymentManagerId] = await findPaymentManagerAddress(name);
    return paymentManagerProgram.instruction.close({
        accounts: {
            paymentManager: paymentManagerId,
            collector: collector || CRANK_KEY,
            closer: wallet.publicKey,
        },
    });
};
