// @cardinal/token-manager/dist/esm/programs/timeInvalidator/accounts.js
import { AnchorProvider, BN, BorshAccountsCoder, Program, } from "@project-serum/anchor";
import { TIME_INVALIDATOR_ADDRESS, TIME_INVALIDATOR_IDL } from "./constants";
// TODO fix types
export const getTimeInvalidator = async (connection, timeInvalidatorId) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const timeInvalidatorProgram = new Program(TIME_INVALIDATOR_IDL, TIME_INVALIDATOR_ADDRESS, provider);
    const parsed = await timeInvalidatorProgram.account.timeInvalidator.fetch(timeInvalidatorId);
    return {
        parsed,
        pubkey: timeInvalidatorId,
    };
};
export const getTimeInvalidators = async (connection, timeInvalidatorIds) => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const timeInvalidatorProgram = new Program(TIME_INVALIDATOR_IDL, TIME_INVALIDATOR_ADDRESS, provider);
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    let timeInvalidators = [];
    try {
        timeInvalidators =
            await timeInvalidatorProgram.account.timeInvalidator.fetchMultiple(timeInvalidatorIds);
    }
    catch (e) {
        console.log(e);
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    return timeInvalidators.map((tm, i) => ({
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        parsed: tm,
        pubkey: timeInvalidatorIds[i],
    }));
};
export const getExpiredTimeInvalidators = async (connection) => {
    const programAccounts = await connection.getProgramAccounts(TIME_INVALIDATOR_ADDRESS);
    const expiredTimeInvalidators = [];
    const coder = new BorshAccountsCoder(TIME_INVALIDATOR_IDL);
    programAccounts.forEach((account) => {
        var _a;
        try {
            const timeInvalidatorData = coder.decode("timeInvalidator", account.account.data);
            if ((_a = timeInvalidatorData.expiration) === null || _a === void 0 ? void 0 : _a.lte(new BN(Date.now() / 1000))) {
                expiredTimeInvalidators.push({
                    ...account,
                    parsed: timeInvalidatorData,
                });
            }
        }
        catch (e) {
            console.log(`Failed to decode time invalidator data`);
        }
    });
    return expiredTimeInvalidators;
};
export const getAllTimeInvalidators = async (connection) => {
    const programAccounts = await connection.getProgramAccounts(TIME_INVALIDATOR_ADDRESS);
    const expiredTimeInvalidators = [];
    const coder = new BorshAccountsCoder(TIME_INVALIDATOR_IDL);
    programAccounts.forEach((account) => {
        try {
            const timeInvalidatorData = coder.decode("timeInvalidator", account.account.data);
            expiredTimeInvalidators.push({
                ...account,
                parsed: timeInvalidatorData,
            });
        }
        catch (e) {
            console.log(`Failed to decode time invalidator data`);
        }
    });
    return expiredTimeInvalidators;
};
