// @cardinal/token-manager/dist/cjs/programs/paymentManager/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.close = exports.managePayment = exports.init = void 0;
const anchor_1 = require("@project-serum/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const tokenManager_1 = require("../tokenManager");
const _1 = require(".");
const pda_1 = require("./pda");
const init = async (connection, wallet, name, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const paymentManagerProgram = new anchor_1.Program(_1.PAYMENT_MANAGER_IDL, _1.PAYMENT_MANAGER_ADDRESS, provider);
    const [paymentManagerId] = await (0, pda_1.findPaymentManagerAddress)(name);
    return [
        paymentManagerProgram.instruction.init({
            name: name,
            feeCollector: params.feeCollector,
            makerFeeBasisPoints: params.makerFeeBasisPoints,
            takerFeeBasisPoints: params.takerFeeBasisPoints,
        }, {
            accounts: {
                paymentManager: paymentManagerId,
                authority: params.authority || wallet.publicKey,
                payer: wallet.publicKey,
                systemProgram: web3_js_1.SystemProgram.programId,
            },
        }),
        paymentManagerId,
    ];
};
exports.init = init;
const managePayment = async (connection, wallet, name, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const paymentManagerProgram = new anchor_1.Program(_1.PAYMENT_MANAGER_IDL, _1.PAYMENT_MANAGER_ADDRESS, provider);
    const [paymentManagerId] = await (0, pda_1.findPaymentManagerAddress)(name);
    return paymentManagerProgram.instruction.managePayment(params.paymentAmount, {
        accounts: {
            paymentManager: paymentManagerId,
            payerTokenAccount: params.payerTokenAccount,
            feeCollectorTokenAccount: params.feeCollectorTokenAccount,
            paymentTokenAccount: params.paymentTokenAccount,
            payer: wallet.publicKey,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        },
    });
};
exports.managePayment = managePayment;
const close = async (connection, wallet, name, collector) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const paymentManagerProgram = new anchor_1.Program(_1.PAYMENT_MANAGER_IDL, _1.PAYMENT_MANAGER_ADDRESS, provider);
    const [paymentManagerId] = await (0, pda_1.findPaymentManagerAddress)(name);
    return paymentManagerProgram.instruction.close({
        accounts: {
            paymentManager: paymentManagerId,
            collector: collector || tokenManager_1.CRANK_KEY,
            closer: wallet.publicKey,
        },
    });
};
exports.close = close;
