// @cardinal/staking/dist/cjs/programs/rewardDistributor/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateRewardDistributor = exports.closeRewardEntry = exports.updateRewardEntry = exports.closeRewardDistributor = exports.claimRewards = exports.initRewardEntry = exports.initRewardDistributor = void 0;
const tslib_1 = require("tslib");
const anchor_1 = require("@project-serum/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const _1 = require(".");
const constants_1 = require("./constants");
const pda_1 = require("./pda");
const initRewardDistributor = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    return rewardDistributorProgram.instruction.initRewardDistributor({
        rewardAmount: params.rewardAmount,
        rewardDurationSeconds: params.rewardDurationSeconds,
        maxSupply: params.maxSupply || null,
        supply: params.supply || null,
        kind: params.kind,
        defaultMultiplier: params.defaultMultiplier || null,
        multiplierDecimals: params.multiplierDecimals || null,
    }, {
        accounts: {
            rewardDistributor: params.rewardDistributorId,
            stakePool: params.stakePoolId,
            rewardMint: params.rewardMintId,
            authority: wallet.publicKey,
            payer: wallet.publicKey,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
        remainingAccounts: params.remainingAccountsForKind,
    });
};
exports.initRewardDistributor = initRewardDistributor;
const initRewardEntry = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    return rewardDistributorProgram.instruction.initRewardEntry({
        accounts: {
            rewardEntry: params.rewardEntryId,
            stakeEntry: params.stakeEntryId,
            rewardDistributor: params.rewardDistributor,
            payer: wallet.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
};
exports.initRewardEntry = initRewardEntry;
const claimRewards = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    const [rewardDistributorId] = yield (0, pda_1.findRewardDistributorId)(params.stakePoolId);
    const [rewardEntryId] = yield (0, pda_1.findRewardEntryId)(rewardDistributorId, params.stakeEntryId);
    return rewardDistributorProgram.instruction.claimRewards({
        accounts: {
            rewardEntry: rewardEntryId,
            rewardDistributor: rewardDistributorId,
            stakeEntry: params.stakeEntryId,
            stakePool: params.stakePoolId,
            rewardMint: params.rewardMintId,
            userRewardMintTokenAccount: params.rewardMintTokenAccountId,
            rewardManager: constants_1.REWARD_MANAGER,
            user: wallet.publicKey,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
        remainingAccounts: params.remainingAccountsForKind,
    });
});
exports.claimRewards = claimRewards;
const closeRewardDistributor = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    const [rewardDistributorId] = yield (0, pda_1.findRewardDistributorId)(params.stakePoolId);
    return rewardDistributorProgram.instruction.closeRewardDistributor({
        accounts: {
            rewardDistributor: rewardDistributorId,
            stakePool: params.stakePoolId,
            rewardMint: params.rewardMintId,
            signer: wallet.publicKey,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        },
        remainingAccounts: params.remainingAccountsForKind,
    });
});
exports.closeRewardDistributor = closeRewardDistributor;
const updateRewardEntry = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    const [rewardDistributorId] = yield (0, pda_1.findRewardDistributorId)(params.stakePoolId);
    const [rewardEntryId] = yield (0, pda_1.findRewardEntryId)(rewardDistributorId, params.stakeEntryId);
    return rewardDistributorProgram.instruction.updateRewardEntry({
        multiplier: params.multiplier,
    }, {
        accounts: {
            rewardEntry: rewardEntryId,
            rewardDistributor: rewardDistributorId,
            authority: wallet.publicKey,
        },
    });
});
exports.updateRewardEntry = updateRewardEntry;
const closeRewardEntry = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    return rewardDistributorProgram.instruction.closeRewardEntry({
        accounts: {
            rewardDistributor: params.rewardDistributorId,
            rewardEntry: params.rewardEntryId,
            authority: wallet.publicKey,
        },
    });
};
exports.closeRewardEntry = closeRewardEntry;
const updateRewardDistributor = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const rewardDistributorProgram = new anchor_1.Program(_1.REWARD_DISTRIBUTOR_IDL, _1.REWARD_DISTRIBUTOR_ADDRESS, provider);
    return rewardDistributorProgram.instruction.updateRewardDistributor({
        defaultMultiplier: params.defaultMultiplier,
        multiplierDecimals: params.multiplierDecimals,
    }, {
        accounts: {
            rewardDistributor: params.rewardDistributorId,
            authority: wallet.publicKey,
        },
    });
};
exports.updateRewardDistributor = updateRewardDistributor;
