// @cardinal/staking/dist/cjs/programs/stakePool/instruction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.closeStakeEntry = exports.closeStakePool = exports.returnReceiptMint = exports.updateTotalStakeSeconds = exports.updateStakePool = exports.unstake = exports.stake = exports.claimReceiptMint = exports.initStakeMint = exports.initStakeEntry = exports.authorizeStakeEntry = exports.initStakePool = exports.initPoolIdentifier = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@cardinal/common");
const tokenManager_1 = require("@cardinal/token-manager/dist/cjs/programs/tokenManager");
const pda_1 = require("@cardinal/token-manager/dist/cjs/programs/tokenManager/pda");
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const anchor_1 = require("@project-serum/anchor");
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const _1 = require(".");
const constants_1 = require("./constants");
const pda_2 = require("./pda");
const utils_1 = require("./utils");
const initPoolIdentifier = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.initIdentifier({
        accounts: {
            identifier: params.identifierId,
            payer: wallet.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
};
exports.initPoolIdentifier = initPoolIdentifier;
const initStakePool = (connection, wallet, params) => {
    var _a;
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.initPool({
        overlayText: params.overlayText,
        imageUri: params.imageUri,
        requiresCollections: params.requiresCollections,
        requiresCreators: params.requiresCreators,
        requiresAuthorization: (_a = params.requiresAuthorization) !== null && _a !== void 0 ? _a : false,
        authority: params.authority,
        resetOnStake: params.resetOnStake,
        cooldownSeconds: params.cooldownSeconds,
        minStakeSeconds: params.minStakeSeconds,
    }, {
        accounts: {
            stakePool: params.stakePoolId,
            identifier: params.identifierId,
            payer: wallet.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
};
exports.initStakePool = initStakePool;
const authorizeStakeEntry = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    const [stakeAuthorizationId] = yield (0, pda_2.findStakeAuthorizationId)(params.stakePoolId, params.originalMintId);
    return stakePoolProgram.instruction.authorizeMint(params.originalMintId, {
        accounts: {
            stakePool: params.stakePoolId,
            stakeAuthorizationRecord: stakeAuthorizationId,
            payer: wallet.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
});
exports.authorizeStakeEntry = authorizeStakeEntry;
const initStakeEntry = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    const remainingAccounts = yield (0, utils_1.remainingAccountsForInitStakeEntry)(params.stakePoolId, params.originalMintId);
    return stakePoolProgram.instruction.initEntry(wallet.publicKey, {
        accounts: {
            stakeEntry: params.stakeEntryId,
            stakePool: params.stakePoolId,
            originalMint: params.originalMintId,
            originalMintMetadata: params.originalMintMetadatId,
            payer: wallet.publicKey,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
        remainingAccounts,
    });
});
exports.initStakeEntry = initStakeEntry;
const initStakeMint = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.initStakeMint({ name: params.name, symbol: params.symbol }, {
        accounts: {
            stakeEntry: params.stakeEntryId,
            stakePool: params.stakePoolId,
            originalMint: params.originalMintId,
            originalMintMetadata: params.originalMintMetadatId,
            stakeMint: params.stakeMintId,
            stakeMintMetadata: params.stakeMintMetadataId,
            stakeEntryStakeMintTokenAccount: params.stakeEntryStakeMintTokenAccountId,
            mintManager: params.mintManagerId,
            payer: wallet.publicKey,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            tokenManagerProgram: tokenManager_1.TOKEN_MANAGER_ADDRESS,
            associatedToken: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
            tokenMetadataProgram: mpl_token_metadata_1.MetadataProgram.PUBKEY,
            systemProgram: web3_js_1.SystemProgram.programId,
        },
    });
};
exports.initStakeMint = initStakeMint;
const claimReceiptMint = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    const [[tokenManagerId], [mintCounterId], stakeEntryReceiptMintTokenAccountId, userReceiptMintTokenAccountId, remainingAccounts,] = yield Promise.all([
        (0, pda_1.findTokenManagerAddress)(params.receiptMintId),
        (0, pda_1.findMintCounterId)(params.receiptMintId),
        (0, common_1.findAta)(params.receiptMintId, params.stakeEntryId, true),
        (0, common_1.findAta)(params.receiptMintId, wallet.publicKey, true),
        (0, tokenManager_1.getRemainingAccountsForKind)(params.receiptMintId, params.receiptType === constants_1.ReceiptType.Original
            ? tokenManager_1.TokenManagerKind.Edition
            : tokenManager_1.TokenManagerKind.Managed),
    ]);
    return stakePoolProgram.instruction.claimReceiptMint({
        accounts: {
            stakeEntry: params.stakeEntryId,
            originalMint: params.originalMintId,
            receiptMint: params.receiptMintId,
            stakeEntryReceiptMintTokenAccount: stakeEntryReceiptMintTokenAccountId,
            user: wallet.publicKey,
            userReceiptMintTokenAccount: userReceiptMintTokenAccountId,
            mintCounter: mintCounterId,
            tokenManager: tokenManagerId,
            tokenManagerReceiptMintTokenAccount: params.tokenManagerReceiptMintTokenAccountId,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            tokenManagerProgram: tokenManager_1.TOKEN_MANAGER_ADDRESS,
            systemProgram: web3_js_1.SystemProgram.programId,
            associatedTokenProgram: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        },
        remainingAccounts,
    });
});
exports.claimReceiptMint = claimReceiptMint;
const stake = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.stake(params.amount, {
        accounts: {
            stakeEntry: params.stakeEntryId,
            stakePool: params.stakePoolId,
            stakeEntryOriginalMintTokenAccount: params.stakeEntryOriginalMintTokenAccountId,
            originalMint: params.originalMint,
            user: wallet.publicKey,
            userOriginalMintTokenAccount: params.userOriginalMintTokenAccountId,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        },
    });
};
exports.stake = stake;
const unstake = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.unstake({
        accounts: {
            stakePool: params.stakePoolId,
            stakeEntry: params.stakeEntryId,
            originalMint: params.originalMintId,
            stakeEntryOriginalMintTokenAccount: params.stakeEntryOriginalMintTokenAccount,
            user: params.user,
            userOriginalMintTokenAccount: params.userOriginalMintTokenAccount,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
        },
        remainingAccounts: params.remainingAccounts,
    });
};
exports.unstake = unstake;
const updateStakePool = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.updatePool({
        overlayText: params.overlayText || "RENTED",
        imageUri: params.imageUri || "",
        requiresCollections: params.requiresCollections || null,
        requiresCreators: params.requiresCreators || null,
        requiresAuthorization: params.requiresAuthorization || false,
        authority: params.authority || null,
        resetOnStake: params.resetOnStake || false,
        cooldownSeconds: params.cooldownSeconds || null,
        minStakeSeconds: params.minStakeSeconds || null,
    }, {
        accounts: {
            stakePool: params.stakePoolId,
            payer: wallet.publicKey,
        },
    });
};
exports.updateStakePool = updateStakePool;
const updateTotalStakeSeconds = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.updateTotalStakeSeconds({
        accounts: {
            stakeEntry: params.stakEntryId,
            lastStaker: params.lastStaker,
        },
    });
};
exports.updateTotalStakeSeconds = updateTotalStakeSeconds;
const returnReceiptMint = (connection, wallet, params) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    const [tokenManagerId] = yield (0, pda_1.findTokenManagerAddress)(params.receiptMint);
    const tokenManagerTokenAccountId = yield (0, common_1.findAta)(params.receiptMint, (yield (0, pda_1.findTokenManagerAddress)(params.receiptMint))[0], true);
    const userReceiptMintTokenAccount = yield (0, common_1.findAta)(params.receiptMint, wallet.publicKey, true);
    const transferAccounts = yield (0, tokenManager_1.getRemainingAccountsForKind)(params.receiptMint, params.tokenManagerKind);
    return stakePoolProgram.instruction.returnReceiptMint({
        accounts: {
            stakeEntry: params.stakeEntry,
            receiptMint: params.receiptMint,
            tokenManager: tokenManagerId,
            tokenManagerTokenAccount: tokenManagerTokenAccountId,
            userReceiptMintTokenAccount: userReceiptMintTokenAccount,
            user: wallet.publicKey,
            collector: tokenManager_1.CRANK_KEY,
            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            tokenManagerProgram: tokenManager_1.TOKEN_MANAGER_ADDRESS,
            rent: web3_js_1.SYSVAR_RENT_PUBKEY,
        },
        remainingAccounts: [
            ...(params.tokenManagerState === tokenManager_1.TokenManagerState.Claimed
                ? transferAccounts
                : []),
            ...params.returnAccounts,
        ],
    });
});
exports.returnReceiptMint = returnReceiptMint;
const closeStakePool = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.closeStakePool({
        accounts: {
            stakePool: params.stakePoolId,
            authority: params.authority,
        },
    });
};
exports.closeStakePool = closeStakePool;
const closeStakeEntry = (connection, wallet, params) => {
    const provider = new anchor_1.AnchorProvider(connection, wallet, {});
    const stakePoolProgram = new anchor_1.Program(_1.STAKE_POOL_IDL, _1.STAKE_POOL_ADDRESS, provider);
    return stakePoolProgram.instruction.closeStakeEntry({
        accounts: {
            stakePool: params.stakePoolId,
            stakeEntry: params.stakeEntryId,
            authority: params.authority,
        },
    });
};
exports.closeStakeEntry = closeStakeEntry;
