// @cardinal/staking/dist/esm/programs/stakePool/transaction.js
import { __awaiter } from "tslib";
import { findAta, tryGetAccount, withFindOrInitAssociatedTokenAccount, } from "@cardinal/common";
import { tokenManager } from "@cardinal/token-manager/dist/cjs/programs";
import { withRemainingAccountsForReturn } from "@cardinal/token-manager/dist/cjs/programs/tokenManager";
import { findMintManagerId, findTokenManagerAddress, tokenManagerAddressFromMint, } from "@cardinal/token-manager/dist/cjs/programs/tokenManager/pda";
import * as metaplex from "@metaplex-foundation/mpl-token-metadata";
import { BN } from "@project-serum/anchor";
import { getMintSupply } from "../../utils";
import { getRewardDistributor } from "../rewardDistributor/accounts";
import { findRewardDistributorId } from "../rewardDistributor/pda";
import { withClaimRewards } from "../rewardDistributor/transaction";
import { getPoolIdentifier, getStakeEntry, getStakePool } from "./accounts";
import { ReceiptType } from "./constants";
import { authorizeStakeEntry, claimReceiptMint, closeStakeEntry, closeStakePool, initPoolIdentifier, initStakeEntry, initStakeMint, initStakePool, returnReceiptMint, stake, unstake, updateStakePool, updateTotalStakeSeconds, } from "./instruction";
import { findIdentifierId, findStakePoolId } from "./pda";
import { findStakeEntryIdFromMint, withRemainingAccountsForUnstake, } from "./utils";
/**
 * Add init pool identifier instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @returns Transaction, public key for the created pool identifier
 */
export const withInitPoolIdentifier = (transaction, connection, wallet) => __awaiter(void 0, void 0, void 0, function* () {
    const [identifierId] = yield findIdentifierId();
    transaction.add(initPoolIdentifier(connection, wallet, {
        identifierId: identifierId,
    }));
    return [transaction, identifierId];
});
export const withInitStakePool = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    const [identifierId] = yield findIdentifierId();
    const identifierData = yield tryGetAccount(() => getPoolIdentifier(connection));
    const identifier = (identifierData === null || identifierData === void 0 ? void 0 : identifierData.parsed.count) || new BN(1);
    if (!identifierData) {
        transaction.add(initPoolIdentifier(connection, wallet, {
            identifierId: identifierId,
        }));
    }
    const [stakePoolId] = yield findStakePoolId(identifier);
    transaction.add(initStakePool(connection, wallet, {
        identifierId: identifierId,
        stakePoolId: stakePoolId,
        requiresCreators: params.requiresCreators || [],
        requiresCollections: params.requiresCollections || [],
        requiresAuthorization: params.requiresAuthorization,
        overlayText: params.overlayText || "",
        imageUri: params.imageUri || "",
        authority: wallet.publicKey,
        resetOnStake: params.resetOnStake || false,
        cooldownSeconds: params.cooldownSeconds || null,
        minStakeSeconds: params.minStakeSeconds || null,
    }));
    return [transaction, stakePoolId];
});
/**
 * Add init stake entry instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param params
 * @returns Transaction, public key for the created stake entry
 */
export const withInitStakeEntry = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    const [[stakeEntryId], originalMintMetadatId] = yield Promise.all([
        findStakeEntryIdFromMint(connection, wallet.publicKey, params.stakePoolId, params.originalMintId),
        metaplex.Metadata.getPDA(params.originalMintId),
    ]);
    transaction.add(yield initStakeEntry(connection, wallet, {
        stakePoolId: params.stakePoolId,
        stakeEntryId: stakeEntryId,
        originalMintId: params.originalMintId,
        originalMintMetadatId: originalMintMetadatId,
    }));
    return [transaction, stakeEntryId];
});
/**
 * Add authorize stake entry instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param params
 * @returns Transaction
 */
export const withAuthorizeStakeEntry = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    transaction.add(yield authorizeStakeEntry(connection, wallet, {
        stakePoolId: params.stakePoolId,
        originalMintId: params.originalMintId,
    }));
    return transaction;
});
/**
 * Add init stake mint instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param params
 * @returns Transaction, keypair of the created stake mint
 */
export const withInitStakeMint = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    const [[mintManagerId], originalMintMetadataId, stakeMintMetadataId] = yield Promise.all([
        findMintManagerId(params.stakeMintKeypair.publicKey),
        metaplex.Metadata.getPDA(params.originalMintId),
        metaplex.Metadata.getPDA(params.stakeMintKeypair.publicKey),
    ]);
    const stakeEntryStakeMintTokenAccountId = yield findAta(params.stakeMintKeypair.publicKey, params.stakeEntryId, true);
    transaction.add(initStakeMint(connection, wallet, {
        stakePoolId: params.stakePoolId,
        stakeEntryId: params.stakeEntryId,
        originalMintId: params.originalMintId,
        originalMintMetadatId: originalMintMetadataId,
        stakeEntryStakeMintTokenAccountId: stakeEntryStakeMintTokenAccountId,
        stakeMintId: params.stakeMintKeypair.publicKey,
        stakeMintMetadataId: stakeMintMetadataId,
        mintManagerId: mintManagerId,
        name: params.name,
        symbol: params.symbol,
    }));
    return [transaction, params.stakeMintKeypair];
});
/**
 * Add claim receipt mint instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param params
 * @returns Transaction
 */
export const withClaimReceiptMint = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    if (params.receiptType === ReceiptType.Original &&
        (yield getMintSupply(connection, params.receiptMintId)).gt(new BN(1))) {
        throw new Error("Fungible staking and locked reecipt type not supported yet");
    }
    const tokenManagerReceiptMintTokenAccountId = yield withFindOrInitAssociatedTokenAccount(transaction, connection, params.receiptMintId, (yield findTokenManagerAddress(params.receiptMintId))[0], wallet.publicKey, true);
    transaction.add(yield claimReceiptMint(connection, wallet, {
        stakeEntryId: params.stakeEntryId,
        tokenManagerReceiptMintTokenAccountId: tokenManagerReceiptMintTokenAccountId,
        originalMintId: params.originalMintId,
        receiptMintId: params.receiptMintId,
        receiptType: params.receiptType,
    }));
    return transaction;
});
/**
 * Add stake instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param params
 * @returns Transaction
 */
export const withStake = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    const [stakeEntryId] = yield findStakeEntryIdFromMint(connection, wallet.publicKey, params.stakePoolId, params.originalMintId);
    const stakeEntryOriginalMintTokenAccountId = yield withFindOrInitAssociatedTokenAccount(transaction, connection, params.originalMintId, stakeEntryId, wallet.publicKey, true);
    transaction.add(stake(connection, wallet, {
        stakeEntryId: stakeEntryId,
        stakePoolId: params.stakePoolId,
        originalMint: params.originalMintId,
        stakeEntryOriginalMintTokenAccountId: stakeEntryOriginalMintTokenAccountId,
        userOriginalMintTokenAccountId: params.userOriginalMintTokenAccountId,
        amount: params.amount || new BN(1),
    }));
    return transaction;
});
/**
 * Add unstake instructions to a transaction
 * @param transaction
 * @param connection
 * @param wallet
 * @param params
 * @returns Transaction
 */
export const withUnstake = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    const [[stakeEntryId], [rewardDistributorId]] = yield Promise.all([
        findStakeEntryIdFromMint(connection, wallet.publicKey, params.stakePoolId, params.originalMintId),
        yield findRewardDistributorId(params.stakePoolId),
    ]);
    const [stakeEntryData, rewardDistributorData] = yield Promise.all([
        tryGetAccount(() => getStakeEntry(connection, stakeEntryId)),
        tryGetAccount(() => getRewardDistributor(connection, rewardDistributorId)),
    ]);
    const stakePoolData = yield getStakePool(connection, params.stakePoolId);
    if ((!stakePoolData.parsed.cooldownSeconds ||
        stakePoolData.parsed.cooldownSeconds === 0 ||
        ((stakeEntryData === null || stakeEntryData === void 0 ? void 0 : stakeEntryData.parsed.cooldownStartSeconds) &&
            Date.now() / 1000 -
                stakeEntryData.parsed.cooldownStartSeconds.toNumber() >=
                stakePoolData.parsed.cooldownSeconds)) &&
        (!stakePoolData.parsed.minStakeSeconds ||
            stakePoolData.parsed.minStakeSeconds === 0 ||
            ((stakeEntryData === null || stakeEntryData === void 0 ? void 0 : stakeEntryData.parsed.lastStakedAt) &&
                Date.now() / 1000 - stakeEntryData.parsed.lastStakedAt.toNumber() >=
                    stakePoolData.parsed.minStakeSeconds))) {
        // return receipt mint if its claimed
        yield withReturnReceiptMint(transaction, connection, wallet, {
            stakeEntryId: stakeEntryId,
        });
    }
    const stakeEntryOriginalMintTokenAccountId = yield withFindOrInitAssociatedTokenAccount(transaction, connection, params.originalMintId, stakeEntryId, wallet.publicKey, true);
    const userOriginalMintTokenAccountId = yield withFindOrInitAssociatedTokenAccount(transaction, connection, params.originalMintId, wallet.publicKey, wallet.publicKey);
    const remainingAccounts = yield withRemainingAccountsForUnstake(transaction, connection, wallet, stakeEntryId, stakeEntryData === null || stakeEntryData === void 0 ? void 0 : stakeEntryData.parsed.stakeMint);
    transaction.add(unstake(connection, wallet, {
        stakePoolId: params.stakePoolId,
        stakeEntryId: stakeEntryId,
        originalMintId: params.originalMintId,
        user: wallet.publicKey,
        stakeEntryOriginalMintTokenAccount: stakeEntryOriginalMintTokenAccountId,
        userOriginalMintTokenAccount: userOriginalMintTokenAccountId,
        remainingAccounts,
    }));
    // claim any rewards deserved
    if (rewardDistributorData) {
        yield withClaimRewards(transaction, connection, wallet, {
            stakePoolId: params.stakePoolId,
            stakeEntryId: stakeEntryId,
        });
    }
    return transaction;
});
export const withUpdateStakePool = (transaction, connection, wallet, params) => {
    transaction.add(updateStakePool(connection, wallet, {
        stakePoolId: params.stakePoolId,
        requiresCreators: params.requiresCreators,
        requiresCollections: params.requiresCollections,
        requiresAuthorization: params.requiresAuthorization,
        overlayText: params.overlayText,
        imageUri: params.imageUri,
        authority: wallet.publicKey,
        resetOnStake: params.resetOnStake,
        cooldownSeconds: params.cooldownSeconds,
        minStakeSeconds: params.minStakeSeconds,
    }));
    return [transaction, params.stakePoolId];
};
export const withUpdateTotalStakeSeconds = (transaction, connection, wallet, params) => {
    transaction.add(updateTotalStakeSeconds(connection, wallet, {
        stakEntryId: params.stakeEntryId,
        lastStaker: params.lastStaker,
    }));
    return transaction;
};
export const withReturnReceiptMint = (transaction, connection, wallet, params) => __awaiter(void 0, void 0, void 0, function* () {
    const stakeEntryData = yield tryGetAccount(() => getStakeEntry(connection, params.stakeEntryId));
    if (!stakeEntryData) {
        throw new Error(`Stake entry ${params.stakeEntryId.toString()} not found`);
    }
    if (!stakeEntryData.parsed.stakeMintClaimed &&
        !stakeEntryData.parsed.originalMintClaimed) {
        console.log("No receipt mint to return");
        return transaction;
    }
    const receiptMint = stakeEntryData.parsed.stakeMint && stakeEntryData.parsed.stakeMintClaimed
        ? stakeEntryData.parsed.stakeMint
        : stakeEntryData.parsed.originalMint;
    const tokenManagerId = yield tokenManagerAddressFromMint(connection, receiptMint);
    const tokenManagerData = yield tryGetAccount(() => tokenManager.accounts.getTokenManager(connection, tokenManagerId));
    if (!tokenManagerData) {
        return transaction;
    }
    const remainingAccountsForReturn = yield withRemainingAccountsForReturn(transaction, connection, wallet, tokenManagerData);
    transaction.add(yield returnReceiptMint(connection, wallet, {
        stakeEntry: params.stakeEntryId,
        receiptMint: receiptMint,
        tokenManagerKind: tokenManagerData.parsed.kind,
        tokenManagerState: tokenManagerData.parsed.state,
        returnAccounts: remainingAccountsForReturn,
    }));
    return transaction;
});
export const withCloseStakePool = (transaction, connection, wallet, params) => {
    transaction.add(closeStakePool(connection, wallet, {
        stakePoolId: params.stakePoolId,
        authority: wallet.publicKey,
    }));
    return transaction;
};
export const withCloseStakeEntry = (transaction, connection, wallet, params) => {
    transaction.add(closeStakeEntry(connection, wallet, {
        stakePoolId: params.stakePoolId,
        stakeEntryId: params.stakeEntryId,
        authority: wallet.publicKey,
    }));
    return transaction;
};
