// @cardinal/staking/dist/esm/programs/stakePool/utils.js
import { __awaiter } from "tslib";
import { withFindOrInitAssociatedTokenAccount } from "@cardinal/common";
import { AnchorProvider, BN, Program } from "@project-serum/anchor";
import { getMintSupply } from "../../utils";
import { REWARD_DISTRIBUTOR_ADDRESS, REWARD_DISTRIBUTOR_IDL, } from "../rewardDistributor";
import { findRewardDistributorId } from "../rewardDistributor/pda";
import { STAKE_POOL_ADDRESS, STAKE_POOL_IDL } from ".";
import { findStakeAuthorizationId, findStakeEntryId } from "./pda";
export const remainingAccountsForInitStakeEntry = (stakePoolId, originalMintId) => __awaiter(void 0, void 0, void 0, function* () {
    const [stakeAuthorizationRecordId] = yield findStakeAuthorizationId(stakePoolId, originalMintId);
    return [
        {
            pubkey: stakeAuthorizationRecordId,
            isSigner: false,
            isWritable: false,
        },
    ];
});
export const withRemainingAccountsForUnstake = (transaction, connection, wallet, stakeEntryId, receiptMint) => __awaiter(void 0, void 0, void 0, function* () {
    if (receiptMint) {
        const stakeEntryReceiptMintTokenAccount = yield withFindOrInitAssociatedTokenAccount(transaction, connection, receiptMint, stakeEntryId, wallet.publicKey, true);
        return [
            {
                pubkey: stakeEntryReceiptMintTokenAccount,
                isSigner: false,
                isWritable: false,
            },
        ];
    }
    else {
        return [];
    }
});
/**
 * Convenience method to find the stake entry id from a mint
 * NOTE: This will lookup the mint on-chain to get the supply
 * @returns
 */
export const findStakeEntryIdFromMint = (connection, wallet, stakePoolId, originalMintId, isFungible) => __awaiter(void 0, void 0, void 0, function* () {
    if (isFungible === undefined) {
        const supply = yield getMintSupply(connection, originalMintId);
        isFungible = supply.gt(new BN(1));
    }
    return findStakeEntryId(wallet, stakePoolId, originalMintId, isFungible);
});
export const getTotalStakeSeconds = (connection, stakeEntryId) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const stakePoolProgram = new Program(STAKE_POOL_IDL, STAKE_POOL_ADDRESS, provider);
    const parsed = yield stakePoolProgram.account.stakeEntry.fetch(stakeEntryId);
    return parsed.totalStakeSeconds;
});
export const getActiveStakeSeconds = (connection, stakeEntryId) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const stakePoolProgram = new Program(STAKE_POOL_IDL, STAKE_POOL_ADDRESS, provider);
    const parsed = yield stakePoolProgram.account.stakeEntry.fetch(stakeEntryId);
    const UTCNow = Math.floor(Date.now() / 1000);
    const lastStakedAt = parsed.lastStakedAt.toNumber() || UTCNow;
    return parsed.lastStaker ? new BN(UTCNow - lastStakedAt) : new BN(0);
});
export const getUnclaimedRewards = (connection, stakePoolId) => __awaiter(void 0, void 0, void 0, function* () {
    var _a;
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const rewardDistributor = new Program(REWARD_DISTRIBUTOR_IDL, REWARD_DISTRIBUTOR_ADDRESS, provider);
    const [rewardDistributorId] = yield findRewardDistributorId(stakePoolId);
    const parsed = yield rewardDistributor.account.rewardDistributor.fetch(rewardDistributorId);
    return parsed.maxSupply
        ? new BN(((_a = parsed.maxSupply) === null || _a === void 0 ? void 0 : _a.toNumber()) - parsed.rewardsIssued.toNumber())
        : new BN(0);
});
export const getClaimedRewards = (connection, stakePoolId) => __awaiter(void 0, void 0, void 0, function* () {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const provider = new AnchorProvider(connection, null, {});
    const rewardDistributor = new Program(REWARD_DISTRIBUTOR_IDL, REWARD_DISTRIBUTOR_ADDRESS, provider);
    const [rewardDistributorId] = yield findRewardDistributorId(stakePoolId);
    const parsed = yield rewardDistributor.account.rewardDistributor.fetch(rewardDistributorId);
    return parsed.rewardsIssued;
});
